src/exoticatechnologies/modifications/exotics/impl/AmmoCraftingSystem.kt

```diff
@@ -0,0 +1,447 @@
+            // First things first, lets grab a random location on the ship
+            val segmentLocation = if (ship.exactBounds != null) {
+                debugLog("generateRandomLocationOnShip()\tsegments size: ${ship.exactBounds.segments.size}")
+                ship.exactBounds.segments.random().p1
+            } else {
+                shipLocation
+            }
+
+            debugLog("generateRandomLocationOnShip()\tsegmentLocation: ${segmentLocation}")
+
+
+            // Just find a point between segmentLocation and location; if they're the same point apply jittering
+            val retVal = if (segmentLocation != shipLocation) {
+                debugLog("generateRandomLocationOnShip()\tsegment location != ship location")
+                // do a random point here
+                val segmentX = segmentLocation.x
+                val segmentY = segmentLocation.y
+
+                debugLog("generateRandomLocationOnShip()\tsegmentX: ${segmentX}, segmentY: ${segmentY}")
+
+                // We are going to get a random point between the ship location (middle of the ship) and this segment X/Y one
+                var diffX = shipLocation.x - segmentLocation.x
+                var diffY = shipLocation.y - segmentLocation.y
+                debugLog("generateRandomLocationOnShip()\tdiffX: ${diffX}, diffY: ${diffY}")
+                // I'm going to abs this, because I don't want to care if the diff is -1300 or not, it's still larger than 1000
+                // And that's the most usual case I'm seeing, diffs that are beyond a -1000
+                val relativeLocation = if (abs(diffX) > 1000 || abs(diffY) > 1000) {
+                    // Even though this could technically be a valid number, more likely this means that
+                    // our ship was at e.g. [0, -1500] and the segment at like [-133, -48] which just means
+                    // there's no way this segment location isn't relative.
+                    true
+                } else { false }
+
+                if (relativeLocation) {
+                    debugLog("based on the diff amount, it was determined that the segmentLocation had to be relative, so diffX/Y are being recalibrated")
+                    diffX = segmentLocation.x
+                    diffY = segmentLocation.y
+                    debugLog("recalibrated diffX: ${diffX}, diffY: ${diffY}")
+                }
+
+                val roundedDiffX = diffX.roundToInt()
+                val roundedDiffY = diffY.roundToInt()
+
+                val scaledDiffX = (3/4f * roundedDiffX).roundToInt()
+                val scaledDiffY = (3/4f * roundedDiffY).roundToInt()
+
+                val randX = MathUtils.getRandomNumberInRange(shipLocation.x, shipLocation.x + scaledDiffX)
+                val randY = MathUtils.getRandomNumberInRange(shipLocation.y, shipLocation.y + scaledDiffY)
```

Comment:
code smell

Reply:
While this particular thing might not be completely replaced, the latest fixes do improve the situation.

Reply:
re-review please. resolving. the smelly part has been deleted, now we just generate a point, and if it's out of bounds we just approximate to the closest point within bounds.